`reconnect` doesn't work

With Simvue client 2.0.0a1 using this test code:
```
from simvue import Run

if __name__ == "__main__":
    with Run() as run:
        run.init(running=False)
        run_id = run.id

    time.sleep(10)

    with Run() as run:
        run.reconnect(run_id)
```
I get:
```
Traceback (most recent call last):
  File "/home/alahiff/client/290250220/reconnet1.py", line 14, in <module>
    run.reconnect(run_id)
  File "/home/alahiff/client/290250220/venv/lib/python3.10/site-packages/simvue/utilities.py", line 245, in wrapper
    return class_func(self, *args, **kwargs)
  File "/home/alahiff/client/290250220/venv/lib/python3.10/site-packages/pydantic/_internal/_validate_call.py", line 38, in wrapper_function
    return wrapper(*args, **kwargs)
  File "/home/alahiff/client/290250220/venv/lib/python3.10/site-packages/pydantic/_internal/_validate_call.py", line 111, in __call__
    res = self.__pydantic_validator__.validate_python(pydantic_core.ArgsKwargs(args, kwargs))
  File "/home/alahiff/client/290250220/venv/lib/python3.10/site-packages/simvue/run.py", line 935, in reconnect
    self._start(reconnect=True)
  File "/home/alahiff/client/290250220/venv/lib/python3.10/site-packages/simvue/run.py", line 471, in _start
    self._sv_obj.status = self._status
  File "/home/alahiff/client/290250220/venv/lib/python3.10/site-packages/simvue/api/objects/base.py", line 67, in _wrapper
    raise AssertionError(
AssertionError: Cannot set property 'status' on read-only object of type 'run'
```

--- a/simvue/client.py
+++ b/simvue/client.py
@@ -990,17 +990,23 @@ class Client:
         RuntimeError
             if there was a failure retrieving data from the server
         """
-
         if not run_id:
+            if critical_only:
+                raise RuntimeError(
+                    "critical_only is ambiguous when returning alerts with no run ID specified."
+                )
             return [alert.name if names_only else alert for _, alert in Alert.get()]  # type: ignore
 
-        return [
-            alert.get("name")
-            if names_only
-            else Alert(identifier=alert.get("id"), **alert)
+        _alerts = [
+            Alert(identifier=alert.get("id"), **alert)
             for alert in Run(identifier=run_id).get_alert_details()
-            if not critical_only or alert["status"].get("current") == "critical"
-        ]  # type: ignore
+        ]
+
+        return [
+            alert.name if names_only else alert
+            for alert in _alerts
+            if not critical_only or alert.get_status(run_id) == "critical"
+        ]
 
     @prettify_pydantic
     @pydantic.validate_call

--- a/simvue/executor.py
+++ b/simvue/executor.py
@@ -348,9 +348,11 @@ class Executor:
                 if self._runner._dispatcher:
                     self._runner._dispatcher.purge()
 
-                self._runner.log_alert(self._alert_ids[proc_id], "critical")
+                self._runner.log_alert(
+                    identifier=self._alert_ids[proc_id], state="critical"
+                )
             else:
-                self._runner.log_alert(self._alert_ids[proc_id], "ok")
+                self._runner.log_alert(identifier=self._alert_ids[proc_id], state="ok")
 
             _current_time: float = 0
             while (

--- a/simvue/metrics.py
+++ b/simvue/metrics.py
@@ -36,14 +36,18 @@ def get_process_memory(processes: list[psutil.Process]) -> int:
     return rss
 
 
-def get_process_cpu(processes: list[psutil.Process]) -> int:
+def get_process_cpu(
+    processes: list[psutil.Process], interval: float | None = None
+) -> int:
     """
     Get the CPU usage
+
+    If first time being called, use a small interval to collect initial CPU metrics.
     """
     cpu_percent: int = 0
     for process in processes:
         with contextlib.suppress(Exception):
-            cpu_percent += process.cpu_percent()
+            cpu_percent += process.cpu_percent(interval=interval)
 
     return cpu_percent
 

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -931,7 +931,7 @@ class Run:
         self._status = "running"
 
         self._id = run_id
-        self._sv_obj = RunObject(identifier=self._id)
+        self._sv_obj = RunObject(identifier=self._id, _read_only=False)
         self._start(reconnect=True)
 
         return True

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -305,15 +305,20 @@ class Run:
 
         return list(set(process_list))
 
-    def _get_sysinfo(self) -> dict[str, typing.Any]:
+    def _get_sysinfo(self, interval: float | None = None) -> dict[str, typing.Any]:
         """Retrieve system administration
+
+        Parameters
+        ----------
+        interval : float | None
+            The interval to use for collection of CPU metrics, by default None (non blocking)
 
         Returns
         -------
         dict[str, typing.Any]
             retrieved system specifications
         """
-        cpu = get_process_cpu(self.processes)
+        cpu = get_process_cpu(self.processes, interval=interval)
         memory = get_process_memory(self.processes)
         gpu = get_gpu_metrics(self.processes)
         data: dict[str, typing.Any] = {}
@@ -347,6 +352,12 @@ class Run:
         ) -> None:
             if not heartbeat_trigger:
                 raise RuntimeError("Expected initialisation of heartbeat")
+
+            # Add initial resource metrics
+            if self._resources_metrics_interval:
+                self._add_metrics_to_dispatch(
+                    self._get_sysinfo(interval=0.1), join_on_fail=False
+                )
 
             last_heartbeat = time.time()
             last_res_metric_call = time.time()

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -1621,7 +1621,6 @@ class Run:
             if description:
                 self._folder.description = description
             self._folder.commit()
-            self._folder.read_only(True)
         except (RuntimeError, ValueError, pydantic.ValidationError) as e:
             self._error(f"Failed to update folder '{self._folder.name}' details: {e}")
             return False

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -958,6 +958,7 @@ class Run:
             PID of the process to be monitored
         """
         self._pid = pid
+        self._parent_process = psutil.Process(self._pid)
 
     @skip_if_failed("_aborted", "_suppress_errors", False)
     @pydantic.validate_call

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -1614,7 +1614,6 @@ class Run:
             return False
 
         try:
-            self._folder.read_only(False)
             if metadata:
                 self._folder.metadata = metadata
             if tags:

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -710,6 +710,7 @@ class Run:
         self._sv_obj.alerts = []
         self._sv_obj.created = time.time()
         self._sv_obj.notifications = notification
+        self._sv_obj._staging["folder_id"] = self._folder.id
 
         if self._status == "running":
             self._sv_obj.system = get_system()

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -1929,16 +1929,21 @@ class Run:
     @check_run_initialised
     @pydantic.validate_call
     def log_alert(
-        self, identifier: str, state: typing.Literal["ok", "critical"]
+        self,
+        identifier: str | None = None,
+        name: str | None = None,
+        state: typing.Literal["ok", "critical"] = "critical",
     ) -> bool:
-        """Set the state of an alert
+        """Set the state of an alert - either specify the alert by ID or name.
 
         Parameters
         ----------
-        identifier : str
-            identifier of alert to update
+        identifier : str | None
+            ID of alert to update, by default None
+        name : str | None
+            Name of the alert to update, by default None
         state : Literal['ok', 'critical']
-            state to set alert to
+            state to set alert to, by default 'critical'
 
         Returns
         -------
@@ -1949,13 +1954,33 @@ class Run:
             self._error('state must be either "ok" or "critical"')
             return False
 
+        if (identifier and name) or (not identifier and not name):
+            self._error("Please specify alert to update either by ID or by name.")
+            return False
+
+        if name:
+            try:
+                if alerts := Alert.get(offline=self._user_config.run.mode == "offline"):
+                    identifier = next(
+                        (id for id, alert in alerts if alert.name == name), None
+                    )
+                else:
+                    self._error("No existing alerts")
+                    return False
+            except RuntimeError as e:
+                self._error(f"{e.args[0]}")
+                return False
+
+        if not identifier:
+            self._error(f"Alert with name '{name}' could not be found.")
+
         _alert = UserAlert(identifier=identifier)
-        # if not isinstance(_alert, UserAlert):
-        #     self._error(
-        #         f"Cannot update state for alert '{identifier}' "
-        #         f"of type '{_alert.__class__.__name__.lower()}'"
-        #     )
-        #     return False
+        if not isinstance(_alert, UserAlert):
+            self._error(
+                f"Cannot update state for alert '{identifier}' "
+                f"of type '{_alert.__class__.__name__.lower()}'"
+            )
+            return False
         _alert.read_only(False)
         _alert.set_status(run_id=self._id, status=state)
         _alert.commit()

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -318,9 +318,10 @@ class Run:
         dict[str, typing.Any]
             retrieved system specifications
         """
-        cpu = get_process_cpu(self.processes, interval=interval)
-        memory = get_process_memory(self.processes)
-        gpu = get_gpu_metrics(self.processes)
+        processes = self.processes
+        cpu = get_process_cpu(processes, interval=0.1)
+        memory = get_process_memory(processes)
+        gpu = get_gpu_metrics(processes)
         data: dict[str, typing.Any] = {}
 
         if memory is not None and cpu is not None:
@@ -352,12 +353,6 @@ class Run:
         ) -> None:
             if not heartbeat_trigger:
                 raise RuntimeError("Expected initialisation of heartbeat")
-
-            # Add initial resource metrics
-            if self._resources_metrics_interval:
-                self._add_metrics_to_dispatch(
-                    self._get_sysinfo(interval=0.1), join_on_fail=False
-                )
 
             last_heartbeat = time.time()
             last_res_metric_call = time.time()

--- a/simvue/run.py
+++ b/simvue/run.py
@@ -305,7 +305,7 @@ class Run:
 
         return list(set(process_list))
 
-    def _get_sysinfo(self, interval: float | None = None) -> dict[str, typing.Any]:
+    def _get_sysinfo(self) -> dict[str, typing.Any]:
         """Retrieve system administration
 
         Parameters
@@ -356,6 +356,9 @@ class Run:
 
             last_heartbeat = time.time()
             last_res_metric_call = time.time()
+
+            if self._resources_metrics_interval:
+                self._add_metrics_to_dispatch(self._get_sysinfo(), join_on_fail=False)
 
             while not heartbeat_trigger.is_set():
                 time.sleep(0.1)

--- a/simvue/sender.py
+++ b/simvue/sender.py
@@ -166,6 +166,8 @@ def sender(
     _user_config = SimvueConfiguration.fetch()
     cache_dir = cache_dir or _user_config.offline.cache
 
+    cache_dir.joinpath("server_ids").mkdir(parents=True, exist_ok=True)
+
     # Check that no other sender is already currently running...
     if cache_dir.joinpath("sender.lock").exists():
         raise RuntimeError("A sender is already running for this cache!")
@@ -173,7 +175,6 @@ def sender(
     # Create lock file to prevent other senders running while this one isn't finished
     cache_dir.joinpath("sender.lock").touch()
 
-    cache_dir.joinpath("server_ids").mkdir(parents=True, exist_ok=True)
     _id_mapping: dict[str, str] = {
         file_path.name.split(".")[0]: file_path.read_text()
         for file_path in cache_dir.glob("server_ids/*.txt")

--- a/simvue/sender.py
+++ b/simvue/sender.py
@@ -92,6 +92,7 @@ def upload_cached_file(
     _logger.info(
         f"{'Updated' if id_mapping.get(_current_id) else 'Created'} {obj_for_upload.__class__.__name__} '{_new_id}'"
     )
+
     file_path.unlink(missing_ok=True)
     if issubclass(_instance_class, simvue.api.objects.ObjectArtifact):
         file_path.parent.joinpath(f"{_current_id}.object").unlink()
@@ -106,9 +107,11 @@ def upload_cached_file(
         obj_type == "runs"
         and cache_dir.joinpath(f"{obj_type}", f"{_current_id}.closed").exists()
     ):
-        # Get list of alerts created by this run - their IDs can be deleted
+        # Get alerts and folder created by this run - their IDs can be deleted
         for id in _data.get("alerts", []):
             cache_dir.joinpath("server_ids", f"{id}.txt").unlink()
+        if _folder_id := _data.get("folder_id"):
+            cache_dir.joinpath("server_ids", f"{_folder_id}.txt").unlink()
 
         cache_dir.joinpath("server_ids", f"{_current_id}.txt").unlink()
         cache_dir.joinpath(f"{obj_type}", f"{_current_id}.closed").unlink()

--- a/simvue/sender.py
+++ b/simvue/sender.py
@@ -170,9 +170,14 @@ def sender(
     }
     _lock = threading.Lock()
     _upload_order = [item for item in UPLOAD_ORDER if item in objects_to_upload]
+    # Glob all files to look in at the start, to prevent extra files being written while other types are being uploaded
+    _all_offline_files = {
+        obj_type: list(cache_dir.glob(f"{obj_type}/*.json"))
+        for obj_type in _upload_order
+    }
 
     for _obj_type in _upload_order:
-        _offline_files = list(cache_dir.glob(f"{_obj_type}/*.json"))
+        _offline_files = _all_offline_files[_obj_type]
         if len(_offline_files) < threading_threshold:
             for file_path in _offline_files:
                 upload_cached_file(cache_dir, _obj_type, file_path, _id_mapping, _lock)

--- a/simvue/sender.py
+++ b/simvue/sender.py
@@ -165,6 +165,14 @@ def sender(
     """
     _user_config = SimvueConfiguration.fetch()
     cache_dir = cache_dir or _user_config.offline.cache
+
+    # Check that no other sender is already currently running...
+    if cache_dir.joinpath("sender.lock").exists():
+        raise RuntimeError("A sender is already running for this cache!")
+
+    # Create lock file to prevent other senders running while this one isn't finished
+    cache_dir.joinpath("sender.lock").touch()
+
     cache_dir.joinpath("server_ids").mkdir(parents=True, exist_ok=True)
     _id_mapping: dict[str, str] = {
         file_path.name.split(".")[0]: file_path.read_text()
@@ -223,4 +231,6 @@ def sender(
                 ),
                 _heartbeat_files,
             )
+    # Remove lock file to allow another sender to start in the future
+    cache_dir.joinpath("sender.lock").unlink()
     return _id_mapping

--- a/simvue/sender.py
+++ b/simvue/sender.py
@@ -116,6 +116,31 @@ def upload_cached_file(
         _logger.info(f"Run {_current_id} closed - deleting cached copies...")
 
 
+def send_heartbeat(
+    file_path: pydantic.FilePath,
+    id_mapping: dict[str, str],
+    server_url: str,
+    headers: dict[str, str],
+):
+    _offline_id = file_path.name.split(".")[0]
+    _online_id = id_mapping.get(_offline_id)
+    if not _online_id:
+        # Run has been closed - can just remove heartbeat and continue
+        file_path.unlink()
+        return
+    _logger.info(f"Sending heartbeat to run {_online_id}")
+    _response = requests.put(
+        f"{server_url}/runs/{_online_id}/heartbeat",
+        headers=headers,
+    )
+    if _response.status_code == 200:
+        file_path.unlink()
+    else:
+        _logger.warning(
+            f"Attempting to send heartbeat to run {_online_id} returned status code {_response.status_code}."
+        )
+
+
 @pydantic.validate_call
 def sender(
     cache_dir: pydantic.DirectoryPath | None = None,
@@ -169,23 +194,26 @@ def sender(
         "Authorization": f"Bearer {_user_config.server.token.get_secret_value()}",
         "User-Agent": f"Simvue Python client {__version__}",
     }
-
-    for _heartbeat_file in cache_dir.glob("runs/*.heartbeat"):
-        _offline_id = _heartbeat_file.name.split(".")[0]
-        _online_id = _id_mapping.get(_offline_id)
-        if not _online_id:
-            # Run has been closed - can just remove heartbeat and continue
-            _heartbeat_file.unlink()
-            continue
-        _logger.info(f"Sending heartbeat to run {_online_id}")
-        _response = requests.put(
-            f"{_user_config.server.url}/runs/{_online_id}/heartbeat",
-            headers=_headers,
-        )
-        if _response.status_code == 200:
-            _heartbeat_file.unlink()
-        else:
-            _logger.warning(
-                f"Attempting to send heartbeat to run {_online_id} returned status code {_response.status_code}."
+    _heartbeat_files = list(cache_dir.glob("runs/*.heartbeat"))
+    if len(_heartbeat_files) < threading_threshold:
+        for _heartbeat_file in _heartbeat_files:
+            (
+                send_heartbeat(
+                    file_path=_heartbeat_file,
+                    id_mapping=_id_mapping,
+                    server_url=_user_config.server.url,
+                    headers=_headers,
+                ),
+            )
+    else:
+        with ThreadPoolExecutor(max_workers=max_workers) as executor:
+            _results = executor.map(
+                lambda _heartbeat_file: send_heartbeat(
+                    file_path=_heartbeat_file,
+                    id_mapping=_id_mapping,
+                    server_url=_user_config.server.url,
+                    headers=_headers,
+                ),
+                _heartbeat_files,
             )
     return _id_mapping

--- a/simvue/sender.py
+++ b/simvue/sender.py
@@ -89,10 +89,9 @@ def upload_cached_file(
         raise RuntimeError(
             f"Object of type '{obj_for_upload.__class__.__name__}' has no identifier"
         )
-    if id_mapping.get(_current_id):
-        _logger.info(f"Updated {obj_for_upload.__class__.__name__} '{_new_id}'")
-    else:
-        _logger.info(f"Created {obj_for_upload.__class__.__name__} '{_new_id}'")
+    _logger.info(
+        f"{'Updated' if id_mapping.get(_current_id) else 'Created'} {obj_for_upload.__class__.__name__} '{_new_id}'"
+    )
     file_path.unlink(missing_ok=True)
     if issubclass(_instance_class, simvue.api.objects.ObjectArtifact):
         file_path.parent.joinpath(f"{_current_id}.object").unlink()

--- a/tests/functional/test_client.py
+++ b/tests/functional/test_client.py
@@ -12,7 +12,7 @@ import simvue.client as svc
 from simvue.exception import ObjectNotFoundError
 import simvue.run as sv_run
 import simvue.api.objects as sv_api_obj
-
+from simvue.api.objects.alert.base import AlertBase
 
 @pytest.mark.dependency
 @pytest.mark.client
@@ -24,25 +24,60 @@ def test_get_events(create_test_run: tuple[sv_run.Run, dict]) -> None:
 @pytest.mark.dependency
 @pytest.mark.client
 @pytest.mark.parametrize(
-    "from_run", (True, False)
+    "from_run", (True, False), ids=("from_run", "all_runs")
 )
-def test_get_alerts(create_test_run: tuple[sv_run.Run, dict], from_run: bool) -> None:
-    time.sleep(1.0)
+@pytest.mark.parametrize(
+    "names_only", (True, False), ids=("names_only", "all_details")
+)
+@pytest.mark.parametrize(
+    "critical_only", (True, False), ids=("critical_only", "all_states")
+)
+def test_get_alerts(create_plain_run: tuple[sv_run.Run, dict], from_run: bool, names_only: bool, critical_only: bool) -> None:
+    run, run_data = create_plain_run
+    run_id = run.id
+    unique_id = f"{uuid.uuid4()}".split("-")[0]
+    _id_1 = run.create_user_alert(
+        name=f"user_alert_1_{unique_id}", 
+    )
+    _id_2 = run.create_user_alert(
+        name=f"user_alert_2_{unique_id}", 
+    )
+    _id_3 = run.create_user_alert(
+        name=f"user_alert_3_{unique_id}",
+        attach_to_run=False
+    )
+    run.log_alert(identifier=_id_1, state="critical")
+    time.sleep(2)
+    run.close()
+    
     client = svc.Client()
-    _, run_data = create_test_run
-    if from_run:
-        triggered_alerts_full = client.get_alerts(run_id=create_test_run[1]["run_id"], critical_only=False, names_only=False)
-        assert len(triggered_alerts_full) == 7
-        for alert in triggered_alerts_full:
-            if alert.name == "value_above_1":
-                assert alert["alert"]["status"]["current"] == "critical"
+    
+    if critical_only and not from_run:
+        with pytest.raises(RuntimeError) as e:
+            _alerts = client.get_alerts(run_id=run_id if from_run else None, critical_only=critical_only, names_only=names_only)
+        assert "critical_only is ambiguous when returning alerts with no run ID specified." in str(e.value)
     else:
-        assert (triggered_alerts_full := client.get_alerts(names_only=True, critical_only=False))
-
-        for alert in run_data["created_alerts"]:
-            assert alert in triggered_alerts_full, f"Alert '{alert}' was not triggered"
-
-
+        _alerts = client.get_alerts(run_id=run_id if from_run else None, critical_only=critical_only, names_only=names_only)
+    
+        if names_only:
+            assert all(isinstance(item, str) for item in _alerts)
+        else:
+            assert all(isinstance(item, AlertBase) for item in _alerts)            
+            _alerts = [alert.name for alert in _alerts]
+            
+        assert f"user_alert_1_{unique_id}" in _alerts
+            
+        if not from_run:
+            assert len(_alerts) > 2
+            assert f"user_alert_3_{unique_id}" in _alerts
+        else:
+            assert f"user_alert_3_{unique_id}" not in _alerts
+            if critical_only:
+                assert len(_alerts) == 1
+            else:
+                assert len(_alerts) == 2
+                assert f"user_alert_2_{unique_id}" in _alerts
+            
 @pytest.mark.dependency
 @pytest.mark.client
 def test_get_run_id_from_name(create_test_run: tuple[sv_run.Run, dict]) -> None:

--- a/tests/functional/test_run_class.py
+++ b/tests/functional/test_run_class.py
@@ -834,11 +834,10 @@ def test_log_alert() -> None:
     
     # Set alert state to OK by ID
     run.log_alert(identifier=_id, state="ok")
-    time.sleep(1)
+    time.sleep(2)
 
     _alert.refresh()
     assert _alert.get_status(_run_id) == "ok"
-    import pdb; pdb.set_trace()
     
     # Check invalid name throws sensible error
     with pytest.raises(RuntimeError) as e:

--- a/tests/functional/test_run_class.py
+++ b/tests/functional/test_run_class.py
@@ -885,7 +885,7 @@ def test_abort_on_alert_raise(create_plain_run: typing.Tuple[sv_run.Run, dict], 
     alert_id = run.create_user_alert("abort_test", trigger_abort=True)
     run.add_process(identifier="forever_long", executable="bash", c="sleep 10")
     time.sleep(2)
-    run.log_alert(alert_id, "critical")
+    run.log_alert(identifier=alert_id, state="critical")
     time.sleep(1)
     _alert = Alert(identifier=alert_id)
     assert _alert.get_status(run.id) == "critical"

--- a/tests/functional/test_run_class.py
+++ b/tests/functional/test_run_class.py
@@ -929,3 +929,41 @@ def test_run_created_with_no_timeout() -> None:
     client = simvue.Client()
     assert client.get_run(run._id)
 
+@pytest.mark.parametrize("mode", ("online", "offline"), ids=("online", "offline"))
+@pytest.mark.run
+def test_reconnect(mode, monkeypatch: pytest.MonkeyPatch) -> None:
+    if mode == "offline":
+        temp_d = tempfile.TemporaryDirectory()
+        monkeypatch.setenv("SIMVUE_OFFLINE_DIRECTORY", temp_d)
+        
+    with simvue.Run(mode=mode) as run:
+        run.init(
+            name="test_reconnect",
+            folder="/simvue_unit_testing",
+            retention_period="2 minutes",
+            timeout=None,
+            running=False
+        )
+        run_id = run.id
+    if mode == "offline":
+        _id_mapping = sv_send.sender(os.environ["SIMVUE_OFFLINE_DIRECTORY"], 2, 10)
+        run_id = _id_mapping.get(run_id)
+        
+    client = simvue.Client()
+    _created_run = client.get_run(run_id)
+    assert _created_run.status == "created"
+    time.sleep(1)
+    
+    with simvue.Run() as run:
+        run.reconnect(run_id)
+        run.log_metrics({"test_metric": 1})
+        run.log_event("Testing!")
+        
+    if mode == "offline":
+        _id_mapping = sv_send.sender(os.environ["SIMVUE_OFFLINE_DIRECTORY"], 2, 10)
+        
+    _reconnected_run = client.get_run(run_id)
+    assert dict(_reconnected_run.metrics)["test_metric"]["last"] == 1
+    assert client.get_events(run_id)[0]["message"] == "Testing!"
+        
+    

--- a/tests/functional/test_run_class.py
+++ b/tests/functional/test_run_class.py
@@ -806,6 +806,50 @@ def test_add_alerts() -> None:
     for _id in _expected_alerts:
         client.delete_alert(_id)
 
+@pytest.mark.run
+def test_log_alert() -> None:
+    _uuid = f"{uuid.uuid4()}".split("-")[0]
+
+    run = sv_run.Run()
+    run.init(
+        name="test_log_alerts",
+        folder="/simvue_unit_tests",
+        retention_period="1 min",
+        tags=["test_add_alerts"],
+        visibility="tenant"
+    )
+    _run_id = run._id
+    # Create a user alert
+    _id = run.create_user_alert(
+        name=f"user_alert_{_uuid}",
+    )
+
+    # Set alert state to critical by name
+    run.log_alert(name=f"user_alert_{_uuid}", state="critical")
+    time.sleep(1)
+    
+    client = sv_cl.Client()
+    _alert = client.get_alerts(run_id=_run_id, critical_only=False, names_only=False)[0]
+    assert _alert.get_status(_run_id) == "critical"
+    
+    # Set alert state to OK by ID
+    run.log_alert(identifier=_id, state="ok")
+    time.sleep(1)
+
+    _alert.refresh()
+    assert _alert.get_status(_run_id) == "ok"
+    import pdb; pdb.set_trace()
+    
+    # Check invalid name throws sensible error
+    with pytest.raises(RuntimeError) as e:
+        run.log_alert(name="fake_name_1234321", state='critical')
+    assert "Alert with name 'fake_name_1234321' could not be found." in str(e.value)
+
+    # Check you cannot specify both ID and name
+    with pytest.raises(RuntimeError) as e:
+        run.log_alert(identifier="myid", name="myname", state='critical')
+    assert "Please specify alert to update either by ID or by name." in str(e.value)
+
 
 @pytest.mark.run
 def test_abort_on_alert_process(mocker: pytest_mock.MockerFixture) -> None:

